src/main/java/core/basesyntax/bookstore/controller/CategoryController.java

```diff
@@ -0,0 +1,180 @@
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/categories")
+@RequiredArgsConstructor
+@Tag(name = "Category management", description = "Endpoints for managing categories")
+public class CategoryController {
+    private static final char PATH_SEPARATOR = '/';
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @Operation(summary = "Get all categories")
+    @ApiResponse(
+            responseCode = "200",
+            description = "Return list of categories"
+    )
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    public List<CategoryDto> getAllCategories() {
+        return categoryService.findAll();
+    }
+
+    @Operation(summary = "Get category by id")
+    @ApiResponses({
+            @ApiResponse(
+                    responseCode = "200",
+                    description = "Found the category",
+                    content = @Content(schema = @Schema(implementation = CategoryDto.class))
+            ),
+            @ApiResponse(
+                    responseCode = "404",
+                    description = "Didn't find the category",
+                    content = @Content(schema = @Schema(implementation = ExceptionDto.class))
+            )
+    })
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/{id}")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getById(id);
+    }
+
+    @Operation(summary = "Create category")
+    @ApiResponses({
+            @ApiResponse(
+                    responseCode = "201",
+                    description = "Created new category",
+                    content = @Content(schema = @Schema(
+                            implementation = CategoryDto.class)
+                    )
+            ),
+            @ApiResponse(
+                    responseCode = "400",
+                    description = "Validation failed",
+                    content = @Content(schema = @Schema(
+                            implementation = ValidationExceptionDto.class
+                    ))
+            )
+    })
+    @PreAuthorize("hasRole('ADMIN')")
+    @PostMapping
+    public ResponseEntity<CategoryDto> createCategory(@Valid
+                                                      @RequestBody
+                                                      CreateCategoryRequestDto requestDto,
+                                                      HttpServletRequest request) {
```

Comment:
```suggestion
) {
```

Reply:
The idea behind 201 Created HTTP status is that resource is created and it should be followed alongside with response header Location which value is URL to newly created resource (in case such URL exists), that's why I need HttpServletRequest object, because without it, I have only URI(/category/{id}) but not URL(host:port/category/{id}) and also that's why I return ResponseEntity object instead of just DTO (it has very suitable static method ResponseEntity.created(URI url).body(Object body) to return